File: src/main/java/frc/robot/PowerCellDetect.py
import numpy as np
import cv2
import math

cap = cv2.VideoCapture(0)


def callback(x):
    global H_low, H_high, S_low, S_high, V_low, V_high, kernel_size, circle_thresh
    H_low = cv2.getTrackbarPos('low H', 'controls')
    H_high = cv2.getTrackbarPos('high H', 'controls')
    S_low = cv2.getTrackbarPos('low S', 'controls')
    S_high = cv2.getTrackbarPos('high S', 'controls')
    V_low = cv2.getTrackbarPos('low V', 'controls')
    V_high = cv2.getTrackbarPos('high V', 'controls')

    kernel_size = cv2.getTrackbarPos('Kernel Size', 'controls')
    circle_thresh = cv2.getTrackbarPos('Circle Threshold', 'controls') / 100


cv2.namedWindow('controls', 2)
cv2.resizeWindow("controls", 550, 10);

H_low = 20
H_high = 45
S_low = 110
S_high = 255
V_low = 80
V_high = 255

blur_size = 5
kernel_size = 5
circle_thresh = 0.12

cv2.createTrackbar('low H', 'controls', H_low, 179, callback)
cv2.createTrackbar('high H', 'controls', H_high, 179, callback)
cv2.createTrackbar('low S', 'controls', S_low, 255, callback)
cv2.createTrackbar('high S', 'controls', S_high, 255, callback)
cv2.createTrackbar('low V', 'controls', V_low, 255, callback)
cv2.createTrackbar('high V', 'controls', V_high, 255, callback)

cv2.createTrackbar('Kernel Size', 'controls', kernel_size, 16, callback)
cv2.createTrackbar('Circle Threshold', 'controls', int(circle_thresh * 100), 99, callback)


def find_circles(image):
    result = image.copy()
    height, width, channels = image.shape
    circles = []

    kernel = np.ones((kernel_size, kernel_size), np.uint8)
    kernel2x = np.ones((kernel_size * 2, kernel_size * 2), np.uint8)

    blur = cv2.GaussianBlur(image, (blur_size, blur_size), 1)
    hsv = cv2.cvtColor(blur, cv2.COLOR_BGR2HSV)

    lower_hsv = np.array([H_low, S_low, V_low])
    upper_hsv = np.array([H_high, S_high, V_high])
    mask = cv2.inRange(hsv, lower_hsv, upper_hsv)
    open_mask = cv2.morphologyEx(mask, cv2.MORPH_OPEN, kernel2x)
    colored_mask = cv2.bitwise_and(blur, blur, mask=open_mask)

    canny = cv2.Canny(open_mask, 0, 255)
    closed_canny = cv2.morphologyEx(canny, cv2.MORPH_CLOSE, kernel)

    contours, hierarchy = cv2.findContours(closed_canny, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_NONE)
    if len(contours) != 0:
        hull_list = []
        for i in range(len(contours)):
            hull = cv2.convexHull(contours[i])
            x, y, w, h = cv2.boundingRect(hull)
            r = (w + h) / 4
            isCircle = True
            accuracy = 0
            for j in range(len(hull)):
                dist = math.sqrt(((x + w / 2) - hull[j][0][0]) ** 2 + ((y + h / 2) - hull[j][0][1]) ** 2)
                accuracy += abs(dist - r) / (r * circle_thresh)
                if (dist * (1 - circle_thresh) > r) or (dist * (1 + circle_thresh) < r):
                    isCircle = False
            accuracy /= len(hull)
            if isCircle:
                circles.append([int((x + w / 2)), int((y + h / 2)), int(r), accuracy])
                hull_list.append(hull)

    cv2.imshow("colored_mask", colored_mask)
    return circles


if cap.isOpened():
    while True:
        ret, img = cap.read()
        frame = cv2.resize(img, (640, 360))
        circle_result = find_circles(frame)
        for circle in circle_result:
            cv2.circle(frame, (circle[0], circle[1]), circle[2], (255, 100, 75), int(circle[2] / 15))
            cv2.circle(frame, (circle[0], circle[1]), 1, (255, 100, 75), int(circle[2] / 15))
            cv2.putText(frame, str(round((1 - circle[3]) * 100)) + "%", (circle[0] + 5, circle[1] + 5), cv2.FONT_HERSHEY_SIMPLEX, circle[2] / 120, (255, 100, 75), int(circle[2] / 60), cv2.LINE_AA)
        cv2.imshow("result", frame)

        k = cv2.waitKey(1) & 0xFF
        if k == ord('q'):
            break

cap.release()
cv2.destroyAllWindows()

File: src/main/java/frc/robot/PowerCellDetect.java
package frc.robot;
import org.opencv.core.*;
import org.opencv.imgproc.*;
import java.util.*;

public class PowerCellDetect {
	private static final int H_LOW = 20;
	private static final int H_HIGH = 45;
	private static final int S_LOW = 145;
	private static final int S_HIGH = 255;
	private static final int V_LOW = 115;
	private static final int V_HIGH = 255;
	private static final int BLUR_SIZE = 5;
	private static final int KERNEL_SIZE = 5;
	private static final int CIRCLE_SPACING = 50;

	public static Mat detect(Mat input) {
		if(input == null || input.empty()) { return null; }

		Mat result = new Mat();
		input.copyTo(result);
		
		int height = input.height();
		int width = input.width();

		Mat blurred = new Mat();
		Imgproc.GaussianBlur(input, blurred, new Size(BLUR_SIZE, BLUR_SIZE), 1);
		Mat hsv = new Mat();
		Imgproc.cvtColor(blurred, hsv, Imgproc.COLOR_BGR2HSV);
		Scalar lower_hsv = new Scalar(H_LOW, S_LOW, V_LOW);
		Scalar upper_hsv = new Scalar(H_HIGH, S_HIGH, V_HIGH);
		Mat mask = new Mat();
		Core.inRange(hsv, lower_hsv, upper_hsv, mask);
		Mat openMask = new Mat();
		Mat element = Imgproc.getStructuringElement(Imgproc.MORPH_RECT, new Size(3, 3));
		for (int i = 0; i < 8; i++) {
			if (i == 0) {
				Imgproc.morphologyEx(mask, openMask, Imgproc.MORPH_OPEN, element);
			} else {
				Imgproc.morphologyEx(openMask, openMask, Imgproc.MORPH_OPEN, element);
			}
		}
		Mat canny = new Mat();
		Imgproc.Canny(openMask, canny, 0, 255);
		element = Imgproc.getStructuringElement(Imgproc.MORPH_RECT, new Size(KERNEL_SIZE, KERNEL_SIZE));
		Mat closedCanny = new Mat();
		Imgproc.morphologyEx(canny, closedCanny, Imgproc.MORPH_CLOSE, element);
		System.out.println(height);
		System.out.println(width);
		Mat contourImg = new Mat(height, width, CvType.CV_8UC1);
		Mat hierarchy = new Mat();
		List<MatOfPoint> contours = new ArrayList<MatOfPoint>();
		Imgproc.findContours(closedCanny, contours, hierarchy, Imgproc.RETR_EXTERNAL, Imgproc.CHAIN_APPROX_NONE);
		if (contours.size() != 0) {
			List<MatOfPoint> hullList = new ArrayList<MatOfPoint>();
			for (MatOfPoint contour : contours) {
				List<Point> list = new ArrayList<Point>();
				MatOfPoint hullPoints = new MatOfPoint();
				MatOfInt hull = new MatOfInt();
				Imgproc.convexHull(contour, hull);
				for(int i=0;i<hull.toArray().length-1;i+=2) {
					list.add(new Point(hull.toArray()[i], hull.toArray()[i+1]));
				}
				hullPoints.fromList(list);
				Rect rect = Imgproc.boundingRect(hullPoints);
				double r = (rect.width + rect.height) / 4.0;
				if ((Math.pow(Math.PI * r, 2)) * 0.9 < Imgproc.contourArea(hullPoints) && ((Math.pow(Math.PI * r, 2)) * 1.1 > Imgproc.contourArea(hullPoints))) {
					if ((rect.width * 0.9 < rect.height) && (rect.width * 1.1 > rect.height)) {
						if (((2 * Math.PI * r) * 0.9 < Imgproc.arcLength(new MatOfPoint2f(hullPoints), true)) && ((2 * Math.PI * r) * 1.1 > Imgproc.arcLength(new MatOfPoint2f(hullPoints), true))) {
							hullList.add(hullPoints);
						}
					}
				}
			}
			Imgproc.drawContours(contourImg, hullList, -1, new Scalar(255, 255, 255), -1);
		}
		Mat circles = new Mat();
		Imgproc.HoughCircles(contourImg, circles, Imgproc.HOUGH_GRADIENT, 3.5, CIRCLE_SPACING);
		if (!circles.empty()) {
			for (int i = 0; i < circles.cols(); i++) {
				double[] xyr = circles.get(0, i);
				Point center = new Point((int)xyr[0], (int)xyr[1]);
				Imgproc.circle(result, center, (int)xyr[2], new Scalar(255, 100, 75), 4);
				Imgproc.circle(result, center, 3, new Scalar(255, 100, 75), 5);
			}
		}
		return contourImg;
	}
}
